content/install-hook-usercss.js

```diff
@@ -1,123 +1,24 @@
-/* global API */
 'use strict';
 
-(() => {
-  // some weird bug in new Chrome: the content script gets injected multiple times
-  if (typeof window.initUsercssInstall === 'function') return;
-  if (!/text\/(css|plain)/.test(document.contentType) ||
-      !/==userstyle==/i.test(document.body.textContent)) {
-    return;
-  }
-  window.initUsercssInstall = () => {};
-
-  orphanCheck();
-
-  const DELAY = 500;
-  const url = location.href;
-  let sourceCode, port, timer;
-
-  chrome.runtime.onConnect.addListener(onConnected);
-  API.openUsercssInstallPage({url})
-    .catch(err => alert(err));
-
-  function onConnected(newPort) {
-    port = newPort;
-    port.onDisconnect.addListener(stop);
-    port.onMessage.addListener(onMessage);
-  }
-
-  function onMessage(msg, port) {
-    switch (msg.method) {
-      case 'getSourceCode':
-        fetchText(url)
-          .then(text => {
-            sourceCode = sourceCode || text;
-            port.postMessage({
-              method: msg.method + 'Response',
-              sourceCode,
-            });
-          })
-          .catch(err => port.postMessage({
-            method: msg.method + 'Response',
-            error: err.message || String(err),
-          }));
-        break;
-
-      case 'liveReloadStart':
-        start();
-        break;
-
-      case 'liveReloadStop':
-        stop();
-        break;
-    }
-  }
-
-  function fetchText(url) {
-    // XHR throws in Chrome 49
-    // FIXME: choose a correct version
-    // https://github.com/openstyles/stylus/issues/560
-    if (getChromeVersion() <= 49) {
-      return fetch(url)
-        .then(r => r.text())
-        .catch(() => fetchTextXHR(url));
-    }
-    return fetchTextXHR(url);
-  }
-
-  function fetchTextXHR(url) {
-    return new Promise((resolve, reject) => {
-      // you can't use fetch in Chrome under 'file:' protocol
-      const xhr = new XMLHttpRequest();
-      xhr.open('GET', url);
-      xhr.addEventListener('load', () => resolve(xhr.responseText));
-      xhr.addEventListener('error', () => reject(xhr));
-      xhr.send();
+// preventing reregistration if reinjected by tabs.executeScript for whatever reason, just in case
+if (typeof self.oldCode !== 'string') {
+  self.oldCode = (document.querySelector('body > pre') || document.body).textContent;
+  chrome.runtime.onConnect.addListener(port => {
+    if (port.name !== 'downloadSelf') return;
+    const read = r => r.status === 200 ? r.text() : Promise.reject(r.status);
+    const postBack = msg => {
+      port.postMessage(msg);
+      self.oldCode = msg.code;
```

Comment:
We can't blindly set `oldCode` to `msg.code` since it may contain `null`. When `oldCode` is set to `null`, the `typeof` check will be broken and the next call of `getData({timer: true})` will always get the code even if the code is not changed.

Reply:
Right.